src/main/java/org/terasology/lost/OnSpawnSystem.java

```diff
@@ -15,26 +15,173 @@
+import org.terasology.math.geom.Vector2f;
+import org.terasology.math.geom.Vector2i;
+import org.terasology.math.geom.Vector3f;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.polyworld.biome.WhittakerBiome;
+import org.terasology.polyworld.biome.BiomeModel;
+import org.terasology.polyworld.graph.Graph;
+import org.terasology.polyworld.graph.GraphFacet;
+import org.terasology.polyworld.graph.GraphFacetProvider;
+import org.terasology.polyworld.biome.WhittakerBiomeModelFacet;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockUri;
+import org.terasology.world.generation.Region;
+import org.terasology.lost.generator.*;
+import org.terasology.math.geom.ImmutableVector2f;
+import org.terasology.structureTemplates.events.SpawnStructureEvent;
+import org.terasology.structureTemplates.util.BlockRegionTransform;
+import org.terasology.world.generation.facets.SurfaceHeightFacet;
+
+import java.util.Set;
 
 @RegisterSystem(RegisterMode.AUTHORITY)
 public class OnSpawnSystem extends BaseComponentSystem {
     @In
     private EntityManager entityManager;
     @In
     private InventoryManager inventoryManager;
+    @In
+    private Console console;
+    @In
+    private AssetManager assetManager;
+    @In
+    private WorldProvider worldProvider;
 
     @ReceiveEvent
     public void onPlayerSpawn(OnPlayerSpawnedEvent event, EntityRef player, InventoryComponent inventory) {
         inventoryManager.giveItem(player, null, entityManager.create("Lost:antrumSabre"));
+        ProgressTrackingComponent progressTrackingComponent = new ProgressTrackingComponent();
+        progressTrackingComponent.addLevel("Grassland", "Lost:stonehengeWipedOutKey");
+        progressTrackingComponent.addLevel("Tundra", "Lost:pyramidBladesOfTruth");
+        progressTrackingComponent.addLevel("Bare", "Lost:pyramidBladesOfTruth");
+        progressTrackingComponent.addLevel("Scorched", "Lost:pyramidBladesOfTruth");
+        progressTrackingComponent.addLevel("Temperate desert", "Lost:pyramidBladesOfTruth");
+        progressTrackingComponent.addLevel("Temperate rain forest", "Lost:templePlasmaOfFire");
+        progressTrackingComponent.addLevel("Temperate deciduous forest", "Lost:templePlasmaOfFire");
+        progressTrackingComponent.addLevel("Subtropical desert", "Lost:pyramidBladesOfTruth");
+        progressTrackingComponent.addLevel("Shrubland", "Lost:stonehengeWipedOutKey");
+        progressTrackingComponent.addLevel("Marsh", "Lost:stonehengeWipedOutKey");
+        progressTrackingComponent.addLevel("Tropical rain forest", "Lost:templePlasmaOfFire");
+        progressTrackingComponent.addLevel("Tropical seasonal forest", "Lost:templePlasmaOfFire");
+        progressTrackingComponent.addLevel("Lakeshore", "Lost:well");
+        progressTrackingComponent.addLevel("Coast", "Lost:well");
+        progressTrackingComponent.addLevel("Beach", "Lost:well");
```

Comment:
Instead of adding multiple levels for the same puzzle because it can appear in multiple biomes, can we design the `ProgressTrackingComponent` such that we can specify multiple biomes for a level in which it can be spawned? :thinking:

Reply:
Done